File: frontend/src/types/conversation.ts
/**
 * Types for the Conversations feature - thread-level conversation data
 */

/**
 * A conversation thread - either a highlight with comments or a journal discussion.
 */
export interface ConversationThread {
  // Thread identification
  threadId: string;
  threadType: 'highlight' | 'journal_discussion';

  // Journal context
  journalId: string;
  journalTitle: string;
  journalAuthorId: string;
  journalAuthorName: string;

  // Thread-specific content (for highlights)
  highlightText: string | null;
  highlightColor: string | null;

  // Activity timestamps
  lastActivity: string;
  createdAt: string;

  // Comment stats
  commentCount: number;

  // Participation tracking
  participants: string[];
  participantIds: string[];

  // Current user's relationship to this thread
  userParticipated: boolean;
  userStarted: boolean;
  userLastSeen: string | null;
  userLastComment: string | null;

  // Unread status
  isUnread: boolean;
  unreadCount: number;
  hasReplyToUser: boolean;

  // Latest comment preview
  latestCommentText: string | null;
  latestCommentAuthor: string | null;
  latestCommentAuthorId: string | null;
  latestCommentTime: string | null;
}

export interface ThreadsResponse {
  threads: ConversationThread[];
  totalUnread: number;
  threadsWithReplies: number;
  totalCount: number;
  hasMore: boolean;
  nextToken: string | null;
}

export interface UnreadCountResponse {
  totalUnread: number;
  threadsWithReplies: number;
  spaceId: string;
}

export interface MarkReadResponse {
  success: boolean;
  journalId: string;
  spaceId: string;
  threadId?: string;
}

export interface GetThreadsOptions {
  limit?: number;
  offset?: number;
  sort?: 'recent' | 'unread' | 'replies';
  type?: 'highlight' | 'journal_discussion';
  filter?: 'all' | 'participated' | 'unread';
  timeFilter?: 'today' | 'week' | 'month';
  search?: string;
}

/**
 * A journal with its conversation threads grouped together.
 * Used for the grouped card view in ConversationsTab.
 */
export interface GroupedJournalConversations {
  journalId: string;
  journalTitle: string;
  journalAuthorId: string;
  journalAuthorName: string;
  lastActivity: string; // Most recent activity across all threads
  totalCommentCount: number;
  totalUnreadCount: number;
  hasReplyToUser: boolean;
  threads: ConversationThread[];
}

// ========== Legacy types (deprecated) ==========

export interface Conversation {
  journalId: string;
  journalTitle: string;
  journalAuthor: string;
  journalAuthorName: string;
  lastActivity: string;
  lastActivityType: 'highlight_comment' | 'journal_comment' | 'highlight';
  lastActivityHighlightId: string | null;
  highlightCount: number;
  highlightCommentCount: number;
  journalCommentCount: number;
  unreadCount: number;
  participants: string[];
  previewText: string | null;
}

export interface ConversationsResponse {
  conversations: Conversation[];
  totalUnread: number;
  nextToken: string | null;
}

export interface GetConversationsOptions {
  limit?: number;
  sort?: 'recent' | 'unread';
}

File: backend/app/models/conversation.py
"""
Data models for Conversations aggregation (Conversations feature).

These models represent the aggregated view of discussions across journals in a space.
Unlike the ReadStatus model, these are computed/aggregated responses, not stored entities.
"""

from typing import List, Optional
from pydantic import BaseModel, Field, ConfigDict


class ConversationModel(BaseModel):
    """Represents a journal with its discussion activity."""
    journal_id: str = Field(alias="journalId")
    journal_title: str = Field(alias="journalTitle")
    journal_author: str = Field(alias="journalAuthor")
    journal_author_name: str = Field(alias="journalAuthorName")
    last_activity: str = Field(alias="lastActivity")
    highlight_count: int = Field(alias="highlightCount")
    highlight_comment_count: int = Field(alias="highlightCommentCount")
    journal_comment_count: int = Field(alias="journalCommentCount")
    unread_count: int = Field(alias="unreadCount")
    participants: List[str] = Field(default_factory=list)
    preview_text: Optional[str] = Field(None, alias="previewText")

    model_config = ConfigDict(populate_by_name=True, by_alias=True)


class ConversationsResponse(BaseModel):
    """Response model for list of conversations in a space."""
    conversations: List[ConversationModel]
    total_unread: int = Field(alias="totalUnread")
    next_token: Optional[str] = Field(None, alias="nextToken")

    model_config = ConfigDict(populate_by_name=True, by_alias=True)


class UnreadCountResponse(BaseModel):
    """Response model for unread count."""
    total_unread: int = Field(alias="totalUnread")
    space_id: str = Field(alias="spaceId")

    model_config = ConfigDict(populate_by_name=True, by_alias=True)


class MarkReadRequest(BaseModel):
    """Request to mark a journal as read."""
    # Optional: mark specific types as read
    mark_highlight_comments: bool = Field(True, alias="markHighlightComments")
    mark_journal_comments: bool = Field(True, alias="markJournalComments")

    model_config = ConfigDict(populate_by_name=True, by_alias=True)


class MarkReadResponse(BaseModel):
    """Response for marking a journal as read."""
    success: bool
    journal_id: str = Field(alias="journalId")
    space_id: str = Field(alias="spaceId")

    model_config = ConfigDict(populate_by_name=True, by_alias=True)
